App crashes during hardware-less boot-up

Tested on: https://github.com/ubopod/ubo_app/actions/runs/13606591776/job/38038859598

If I attempt to boot up ubo app on bare Raspberry Pi with no hardware add-on, it crashes due to sensor initialization since there are no devices on I2C bus (no keypad, no sensors, etc)

temperature sensor error:

```
1733061055.564713 [ERROR] Error initializing temperature sensor
Traceback (most recent call last):
  File "/opt/ubo/env/lib/python3.11/site-packages/adafruit_bus_device/i2c_device.py", line 175, in __probe_for_device
    self.i2c.writeto(self.device_address, b"")
  File "/opt/ubo/env/lib/python3.11/site-packages/busio.py", line 215, in writeto
    return self._i2c.writeto(address, buffer, stop=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/adafruit_blinka/microcontroller/generic_linux/i2c.py", line 60, in writeto
    self._i2c_bus.write_bytes(address, buffer[start:end])
  File "/opt/ubo/env/lib/python3.11/site-packages/Adafruit_PureIO/smbus.py", line 303, in write_bytes
    self._device.write(buf)
OSError: [Errno 121] Remote I/O error

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/ubo/env/lib/python3.11/site-packages/adafruit_bus_device/i2c_device.py", line 181, in __probe_for_device
    self.i2c.readfrom_into(self.device_address, result)
  File "/opt/ubo/env/lib/python3.11/site-packages/busio.py", line 205, in readfrom_into
    return self._i2c.readfrom_into(address, buffer, stop=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/adafruit_blinka/microcontroller/generic_linux/i2c.py", line 67, in readfrom_into
    readin = self._i2c_bus.read_bytes(address, end - start)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/Adafruit_PureIO/smbus.py", line 170, in read_bytes
    return self._device.read(number)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: [Errno 5] Input/output error

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/ubo/env/lib/python3.11/site-packages/ubo_app/services/040-sensors/setup.py", line 67, in init_service
    temperature_sensor = _initialize_device(adafruit_pct2075.PCT2075, 0x48, i2c)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/tenacity/__init__.py", line 336, in wrapped_f
    return copy(f, *args, **kw)
           ^^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/tenacity/__init__.py", line 475, in __call__
    do = self.iter(retry_state=retry_state)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/tenacity/__init__.py", line 376, in iter
    result = action(retry_state)
             ^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/tenacity/__init__.py", line 398, in <lambda>
    self._add_action_func(lambda rs: rs.outcome.result())
                                     ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/opt/ubo/env/lib/python3.11/site-packages/tenacity/__init__.py", line 478, in __call__
    result = fn(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/ubo_app/services/040-sensors/setup.py", line 56, in _initialize_device
    return cls(i2c, address)
           ^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/adafruit_pct2075.py", line 109, in __init__
    self.i2c_device = i2cdevice.I2CDevice(i2c_bus, address)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/adafruit_bus_device/i2c_device.py", line 62, in __init__
    self.__probe_for_device()
  File "/opt/ubo/env/lib/python3.11/site-packages/adafruit_bus_device/i2c_device.py", line 184, in __probe_for_device
    raise ValueError("No I2C device at address: 0x%x" % self.device_address)
ValueError: No I2C device at address: 0x48
```

Light sensor initialization failing:

```
1733061055.665822 [ERROR] Error initializing light sensor
Traceback (most recent call last):
  File "/opt/ubo/env/lib/python3.11/site-packages/adafruit_bus_device/i2c_device.py", line 175, in __probe_for_device
    self.i2c.writeto(self.device_address, b"")
  File "/opt/ubo/env/lib/python3.11/site-packages/busio.py", line 215, in writeto
    return self._i2c.writeto(address, buffer, stop=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/adafruit_blinka/microcontroller/generic_linux/i2c.py", line 60, in writeto
    self._i2c_bus.write_bytes(address, buffer[start:end])
  File "/opt/ubo/env/lib/python3.11/site-packages/Adafruit_PureIO/smbus.py", line 303, in write_bytes
    self._device.write(buf)
OSError: [Errno 121] Remote I/O error

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/ubo/env/lib/python3.11/site-packages/adafruit_bus_device/i2c_device.py", line 181, in __probe_for_device
    self.i2c.readfrom_into(self.device_address, result)
  File "/opt/ubo/env/lib/python3.11/site-packages/busio.py", line 205, in readfrom_into
    return self._i2c.readfrom_into(address, buffer, stop=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/adafruit_blinka/microcontroller/generic_linux/i2c.py", line 67, in readfrom_into
    readin = self._i2c_bus.read_bytes(address, end - start)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/Adafruit_PureIO/smbus.py", line 170, in read_bytes
    return self._device.read(number)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
OSError: [Errno 5] Input/output error

During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/opt/ubo/env/lib/python3.11/site-packages/ubo_app/services/040-sensors/setup.py", line 72, in init_service
    light_sensor = _initialize_device(adafruit_veml7700.VEML7700, 0x10, i2c)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/tenacity/__init__.py", line 336, in wrapped_f
    return copy(f, *args, **kw)
           ^^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/tenacity/__init__.py", line 475, in call
    do = self.iter(retry_state=retry_state)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/tenacity/__init__.py", line 376, in iter
    result = action(retry_state)
             ^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/tenacity/__init__.py", line 398, in <lambda>
    self._add_action_func(lambda rs: rs.outcome.result())
                                     ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/concurrent/futures/_base.py", line 449, in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
  File "/opt/ubo/env/lib/python3.11/site-packages/tenacity/__init__.py", line 478, in call
    result = fn(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/ubo_app/services/040-sensors/setup.py", line 56, in _initialize_device
    return cls(i2c, address)
           ^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/adafruit_veml7700.py", line 196, in init
    self.i2c_device = i2cdevice.I2CDevice(i2c_bus, address)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/adafruit_bus_device/i2c_device.py", line 62, in init
    self.__probe_for_device()
  File "/opt/ubo/env/lib/python3.11/site-packages/adafruit_bus_device/i2c_device.py", line 184, in __probe_for_device
    raise ValueError("No I2C device at address: 0x%x" % self.device_address)
ValueError: No I2C device at address: 0x10
1733061055.855031 [ERROR] Thread exception
Traceback (most recent call last):
  File "/usr/lib/python3.11/threading.py", line 1038, in _bootstrap_inner
    self.run()
  File "/opt/ubo/env/lib/python3.11/site-packages/sentry_sdk/integrations/threading.py", line 101, in run
    return _run_old_run_func()
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/sentry_sdk/integrations/threading.py", line 96, in _run_old_run_func
    reraise(*_capture_exception())
  File "/opt/ubo/env/lib/python3.11/site-packages/sentry_sdk/utils.py", line 1718, in reraise
    raise value
  File "/opt/ubo/env/lib/python3.11/site-packages/sentry_sdk/integrations/threading.py", line 94, in _run_old_run_func
    return old_run_func(self, *a, **kw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/ubo_app/load_services.py", line 302, in run
    result = cast(Callable[[UboServiceThread], None], self.setup)(self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/ubo/env/lib/python3.11/site-packages/ubo_app/services/040-sensors/ubo_handle.py", line 16, in setup
    init_service()
  File "/opt/ubo/env/lib/python3.11/site-packages/ubo_app/services/040-sensors/setup.py", line 78, in init_service
    read_sensors()
  File "/opt/ubo/env/lib/python3.11/site-packages/ubo_app/services/040-sensors/setup.py", line 28, in read_sensors
    temperature = temperature_sensor.temperature
                  ^^^^^^^^^^^^^^^^^^
NameError: name 'temperature_sensor' is not defined - extra: {
  "exception_thread": "<UboServiceThread id=sensors label=Sensors name=040-sensors>"
}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -20,6 +20,7 @@
 - refactor(voice): upgrade `access_key` input to use the new `fields` value of the `InputDescription`
 - refactor(web-ui): use the same mui interface in static server too (instead of jinja-rendered interface) and add action buttons for all status reported by server - closes #224
 - fix(rpi-connect): use `with_state` of the latest python-redux version instead of `view` to avoid memoization of actions - closes #248
+- refactor(core,audio,display,keypad,sensors): read and check content of eeprom to determine if a device/service should be initialized or not - closes #223, #closes #249
 
 ## Version 1.2.2
 

--- a/ubo_app/display.py
+++ b/ubo_app/display.py
@@ -16,6 +16,7 @@ from ubo_app.store.services.display import (
     DisplayRenderEvent,
 )
 from ubo_app.utils import IS_RPI
+from ubo_app.utils.eeprom import get_eeprom_data
 
 if TYPE_CHECKING:
     from headless_kivy.config import Region
@@ -24,24 +25,34 @@ if TYPE_CHECKING:
 from ubo_app.constants import HEIGHT, WIDTH
 
 if IS_RPI:
-    import board
-    import digitalio
+    eeprom_data = get_eeprom_data()
 
-    cs_pin = digitalio.DigitalInOut(board.CE0)
-    dc_pin = digitalio.DigitalInOut(board.D25)
-    reset_pin = digitalio.DigitalInOut(board.D24)
-    spi = board.SPI()
-    display = ST7789(
-        spi,
-        height=HEIGHT,
-        width=WIDTH,
-        y_offset=80,
-        x_offset=0,
-        cs=cs_pin,
-        dc=dc_pin,
-        rst=reset_pin,
-        baudrate=70_000_000,
-    )
+    if (
+        eeprom_data is not None
+        and 'lcd' in eeprom_data
+        and eeprom_data['lcd'] is not None
+        and eeprom_data['lcd']['model'] == 'st7789'
+    ):
+        import board
+        import digitalio
+
+        cs_pin = digitalio.DigitalInOut(board.CE0)
+        dc_pin = digitalio.DigitalInOut(board.D25)
+        reset_pin = digitalio.DigitalInOut(board.D24)
+        spi = board.SPI()
+        display = ST7789(
+            spi,
+            height=HEIGHT,
+            width=WIDTH,
+            y_offset=80,
+            x_offset=0,
+            cs=cs_pin,
+            dc=dc_pin,
+            rst=reset_pin,
+            baudrate=70_000_000,
+        )
+    else:
+        display = None
 else:
     display = cast(ST7789, Fake())
 
@@ -96,7 +107,7 @@ def render_on_display(*, regions: list[Region]) -> None:
     )
 
 
-original_block = display._block  # noqa: SLF001
+original_block = display._block if display is not None else None  # noqa: SLF001
 
 
 @store.with_state(
@@ -110,19 +121,20 @@ def render_block(
     bypass_pause: bool = False,
 ) -> None:
     """Block the display."""
-    if not is_paused or bypass_pause:
+    if display is not None and (not is_paused or bypass_pause):
         display._block(*rectangle, data_bytes)  # noqa: SLF001
 
 
 def turn_off() -> None:
     """Turn off the display."""
-    display._block = lambda *args, **kwargs: (args, kwargs)  # noqa: SLF001
+    if display is not None:
+        display._block = lambda *args, **kwargs: (args, kwargs)  # noqa: SLF001
     render_blank()
 
 
 def render_blank() -> None:
     """Render a blank screen."""
-    if IS_RPI:
+    if IS_RPI and original_block is not None:
         original_block(0, 0, WIDTH - 1, HEIGHT - 1, b'\x00\x00' * WIDTH * HEIGHT)
         time.sleep(0.2)
         original_block(0, 0, WIDTH - 1, HEIGHT - 1, b'\x00\x00' * WIDTH * HEIGHT)

--- a/ubo_app/services/000-audio/audio_manager.py
+++ b/ubo_app/services/000-audio/audio_manager.py
@@ -15,6 +15,7 @@ from ubo_app.logger import logger
 from ubo_app.store.main import store
 from ubo_app.store.services.audio import AudioPlaybackDoneAction
 from ubo_app.utils.async_ import create_task
+from ubo_app.utils.eeprom import get_eeprom_data
 from ubo_app.utils.server import send_command
 
 CHUNK_SIZE = 1024
@@ -38,13 +39,15 @@ class AudioManager:
     def __init__(self: AudioManager) -> None:
         """Initialize the audio manager."""
         # create an audio object
-        self.cardindex = None
+        self.card_index = None
+        self.has_speakers = False
+        self.has_microphones = False
 
         async def initialize_audio() -> None:
             for _ in range(20):
                 try:
                     cards = alsaaudio.cards()
-                    self.cardindex = cards.index(
+                    self.card_index = cards.index(
                         next(card for card in cards if 'wm8960' in card),
                     )
                 except StopIteration:
@@ -53,7 +56,24 @@ class AudioManager:
                     break
                 await asyncio.sleep(1)
 
-        create_task(initialize_audio())
+        eeprom_data = get_eeprom_data()
+
+        if eeprom_data is not None and (
+            'speakers' in eeprom_data
+            and eeprom_data['speakers']
+            and eeprom_data['speakers']['model'] == 'wm8960'
+        ):
+            self.has_speakers = True
+
+        if eeprom_data is not None and (
+            'microphones' in eeprom_data
+            and eeprom_data['microphones']
+            and eeprom_data['microphones']['model'] == 'wm8960'
+        ):
+            self.has_microphones = True
+
+        if self.has_speakers or self.has_microphones:
+            create_task(initialize_audio())
 
     async def play_file(self: AudioManager, filename: str) -> None:
         """Play a waveform audio file.
@@ -154,17 +174,17 @@ class AudioManager:
             mixer.setmute(1 if mute else 0)
         except alsaaudio.ALSAAudioError:
             # Seems like pulseaudio is not installed, so we directly use device mixers
-            if self.cardindex is None:
+            if self.card_index is None or not self.has_speakers:
                 return
 
             mixer = alsaaudio.Mixer(
                 control='Right Output Mixer PCM',
-                cardindex=self.cardindex,
+                cardindex=self.card_index,
             )
             mixer.setmute(0)
             mixer = alsaaudio.Mixer(
                 control='Left Output Mixer PCM',
-                cardindex=self.cardindex,
+                cardindex=self.card_index,
             )
             mixer.setmute(0)
 
@@ -186,16 +206,16 @@ class AudioManager:
             mixer.setvolume(round(volume * 100))
         except alsaaudio.ALSAAudioError:
             # Seems like pulseaudio is not installed, so we directly use device mixers
-            if self.cardindex is None:
+            if self.card_index is None or not self.has_speakers:
                 return
 
-            mixer = alsaaudio.Mixer(control='Speaker', cardindex=self.cardindex)
+            mixer = alsaaudio.Mixer(control='Speaker', cardindex=self.card_index)
             mixer.setvolume(
                 _linear_to_logarithmic(volume),
                 alsaaudio.MIXER_CHANNEL_ALL,
                 alsaaudio.PCM_PLAYBACK,
             )
-            mixer = alsaaudio.Mixer(control='Playback', cardindex=self.cardindex)
+            mixer = alsaaudio.Mixer(control='Playback', cardindex=self.card_index)
             mixer.setvolume(
                 100,
                 alsaaudio.MIXER_CHANNEL_ALL,
@@ -220,10 +240,10 @@ class AudioManager:
             mixer.setrec(round(volume * 100))
         except alsaaudio.ALSAAudioError:
             # Seems like pulseaudio is not installed, so we directly use device mixers
-            if self.cardindex is None:
+            if self.card_index is None or not self.has_microphones:
                 return
 
-            mixer = alsaaudio.Mixer(control='Capture', cardindex=self.cardindex)
+            mixer = alsaaudio.Mixer(control='Capture', cardindex=self.card_index)
             mixer.setvolume(
                 _linear_to_logarithmic(volume),
                 alsaaudio.MIXER_CHANNEL_ALL,

--- a/ubo_app/services/000-keypad/setup.py
+++ b/ubo_app/services/000-keypad/setup.py
@@ -17,6 +17,7 @@ from ubo_app.store.services.keypad import (
     KeypadKeyReleaseAction,
 )
 from ubo_app.utils import IS_RPI
+from ubo_app.utils.eeprom import get_eeprom_data
 
 if TYPE_CHECKING:
     import busio
@@ -279,4 +280,11 @@ class Keypad:
 def init_service() -> None:
     if not IS_RPI:
         return
-    Keypad()
+    eeprom_data = get_eeprom_data()
+    if (
+        eeprom_data is not None
+        and 'keypad' in eeprom_data
+        and eeprom_data['keypad']
+        and eeprom_data['keypad']['model'] == 'aw9523'
+    ):
+        Keypad()

--- a/ubo_app/services/040-rgb-ring/setup.py
+++ b/ubo_app/services/040-rgb-ring/setup.py
@@ -1,9 +1,20 @@
 # ruff: noqa: D100, D101, D102, D103, D104, D107, N999
 from ubo_app.store.main import store
 from ubo_app.store.services.rgb_ring import RgbRingCommandEvent, RgbRingPulseAction
+from ubo_app.utils.eeprom import get_eeprom_data
 
 
 def init_service() -> None:
+    eeprom_data = get_eeprom_data()
+
+    if (
+        eeprom_data is None
+        or 'led' not in eeprom_data
+        or eeprom_data['led'] is None
+        or eeprom_data['led']['model'] != 'neopixel'
+    ):
+        return
+
     from rgb_ring_client import RgbRingClient
 
     rgb_ring_client = RgbRingClient()

--- a/ubo_app/services/040-sensors/setup.py
+++ b/ubo_app/services/040-sensors/setup.py
@@ -15,18 +15,19 @@ from tenacity import retry, retry_if_exception, stop_after_attempt, wait_fixed
 from ubo_app.logger import logger
 from ubo_app.store.main import store
 from ubo_app.store.services.sensors import Sensor, SensorsReportReadingAction
+from ubo_app.utils.eeprom import get_eeprom_data
 
 if TYPE_CHECKING:
     from adafruit_rgb_display.rgb import busio
 
-temperature_sensor: adafruit_pct2075.PCT2075
-light_sensor: adafruit_veml7700.VEML7700
+temperature_sensor: adafruit_pct2075.PCT2075 | None = None
+light_sensor: adafruit_veml7700.VEML7700 | None = None
 
 
 def read_sensors(_: float | None = None) -> None:
     """Read the sensor."""
-    temperature = temperature_sensor.temperature
-    light = light_sensor.lux
+    temperature = 0.0 if temperature_sensor is None else temperature_sensor.temperature
+    light = 0.0 if light_sensor is None else light_sensor.lux
     store.dispatch(
         SensorsReportReadingAction(
             sensor=Sensor.TEMPERATURE,
@@ -60,16 +61,38 @@ def init_service() -> None:
     """Initialize the service."""
     from kivy.clock import Clock
 
+    eeprom_data = get_eeprom_data()
+
     global temperature_sensor, light_sensor  # noqa: PLW0603
 
     i2c = board.I2C()
     try:
-        temperature_sensor = _initialize_device(adafruit_pct2075.PCT2075, 0x48, i2c)
+        if (
+            eeprom_data is not None
+            and 'temperature' in eeprom_data
+            and eeprom_data['temperature'] is not None
+            and eeprom_data['temperature']['model'].upper() == 'PCT2075'
+        ):
+            temperature_sensor = _initialize_device(
+                adafruit_pct2075.PCT2075,
+                int(eeprom_data['temperature']['bus_address'], 16),
+                i2c,
+            )
     except Exception:
         logger.exception('Error initializing temperature sensor')
 
     try:
-        light_sensor = _initialize_device(adafruit_veml7700.VEML7700, 0x10, i2c)
+        if (
+            eeprom_data is not None
+            and 'ambient' in eeprom_data
+            and eeprom_data['ambient'] is not None
+            and eeprom_data['ambient']['model'].upper() == 'VEML7700'
+        ):
+            light_sensor = _initialize_device(
+                adafruit_veml7700.VEML7700,
+                int(eeprom_data['ambient']['bus_address'], 16),
+                i2c,
+            )
     except Exception:
         logger.exception('Error initializing light sensor')
 

--- a/ubo_app/store/update_manager/types.py
+++ b/ubo_app/store/update_manager/types.py
@@ -18,7 +18,7 @@ class UpdateManagerSetVersionsAction(UpdateManagerAction):
     current_version: str
     base_image_variant: str
     latest_version: str
-    serial_number: str
+    serial_number: str | None
 
 
 class UpdateManagerSetStatusAction(UpdateManagerAction):

--- a/ubo_app/utils/eeprom.py
+++ b/ubo_app/utils/eeprom.py
@@ -1,21 +1,177 @@
 """Utility functions for interacting with the EEPROM."""
 
 import json
+from functools import cache
 from pathlib import Path
+from typing import TypedDict
 
 from ubo_app.logger import logger
 from ubo_app.utils import IS_RPI
 
 
-def read_serial_number() -> str:
+class SoundTestReport(TypedDict):
+    """Sound test report."""
+
+    left: bool
+    right: bool
+
+
+class TemperatureTestReport(TypedDict):
+    """Temperature test report."""
+
+    degrees: float
+
+
+class AmbientTestReport(TypedDict):
+    """Ambient test report."""
+
+    works: bool
+    baseline: float
+    reading: float
+    delta: float
+
+
+class KeypadTestReport(TypedDict):
+    """Keypad test report."""
+
+    L1: bool
+    L2: bool
+    L3: bool
+    UP: bool
+    DOWN: bool
+    BACK: bool
+    HOME: bool
+    MIC: bool
+
+
+class LCDTestReport(TypedDict):
+    """LCD test report."""
+
+    qrcode: bool
+    green: bool
+    red: bool
+    blue: bool
+
+
+class LEDTestReport(TypedDict):
+    """LED test report."""
+
+    green: bool
+    red: bool
+    blue: bool
+
+
+class Device(TypedDict):
+    """Device information."""
+
+    model: str
+    bus_address: str
+    test_result: bool
+
+
+class EepromDevice(Device):
+    """EEPROM device information."""
+
+
+class SpeakersDevice(Device):
+    """Speakers device information."""
+
+    test_report: SoundTestReport
+
+
+class MicrophonesDevice(Device):
+    """Microphones device information."""
+
+    test_report: SoundTestReport
+
+
+class TemperatureDevice(Device):
+    """Temperature device information."""
+
+    test_report: TemperatureTestReport
+
+
+class AmbientDevice(Device):
+    """Ambient device information."""
+
+    test_report: AmbientTestReport
+
+
+class KeypadDevice(Device):
+    """Keypad device information."""
+
+    test_report: KeypadTestReport
+
+
+class LCDDevice(Device):
+    """LCD device information."""
+
+    test_report: LCDTestReport
+
+
+class LEDDevice(Device):
+    """LED device information."""
+
+    test_report: LEDTestReport
+
+
+class InfraredDevice(Device):
+    """Infrared device information."""
+
+    test_result: bool
+
+
+class I2CBus(TypedDict):
+    """I2C bus information."""
+
+    num_devices: int
+    scanned_addressed: list[str]
+    status: str
+
+
+class EepromData(TypedDict):
+    """EEPROM data."""
+
+    serial_number: str
+    eeprom: EepromDevice | None
+    speakers: SpeakersDevice | None
+    microphones: MicrophonesDevice | None
+    temperature: TemperatureDevice | None
+    ambient: AmbientDevice | None
+    keypad: KeypadDevice | None
+    i2c_bus: I2CBus | None
+    lcd: LCDDevice | None
+    led: LEDDevice | None
+    infrared: InfraredDevice | None
+    version: str
+    timedate: str | None
+    test_result: bool | None
+
+
+@cache
+def get_eeprom_data() -> EepromData | None:
+    """Read the EEPROM data."""
+    if not IS_RPI:
+        return None
+    try:
+        eeprom_json_data = Path(
+            '/proc/device-tree/hat/custom_0',
+        ).read_text(encoding='utf-8')
+        eeprom_data = json.loads(eeprom_json_data)
+    except Exception:
+        logger.exception('Failed to read EEPROM data')
+        return None
+    else:
+        if 'serial_number' not in eeprom_data or 'version' not in eeprom_data:
+            return None
+        return eeprom_data
+
+
+def read_serial_number() -> str | None:
     """Read the serial number from the EEPROM."""
     if IS_RPI:
-        try:
-            eeprom_json_data = Path(
-                '/proc/device-tree/hat/custom_0',
-            ).read_text(encoding='utf-8')
-            eeprom_data = json.loads(eeprom_json_data)
-            return eeprom_data['serial_number']
-        except Exception:
-            logger.exception('Failed to read serial number')
+        data = get_eeprom_data()
+        if data is not None:
+            return data['serial_number']
+        return None
     return '<not-available>'

